src/core_modules/capture-core/src/components/QuickSelector/QuickSelector.component.js

```diff
@@ -17,73 +21,128 @@ const styles = () => ({
+            for (let i = 0; i < programsArray.length; i++) {
+                if (programsArray[i].id === this.props.selectedProgramId) {
+                    selectedProgram = programsArray[i];
+                }
+            }
+        }
+
+        if (this.props.selectedOrgUnitId && this.props.selectedProgramId) {
+            if (selectedProgram.categoryCombo && selectedProgram.categoryCombo.categories.length > 0 && !selectedProgram.categoryCombo.isDefault) {
+                orgUnitSelectorWidth = 3;
+                programSelectorWidth = 5;
+                actionButtonsWidth = 4;
+            } else {
+                orgUnitSelectorWidth = 3;
+                programSelectorWidth = 3;
+                actionButtonsWidth = 6;
+            }
+        } else if (!this.props.selectedOrgUnitId && this.props.selectedProgramId) {
+            if (selectedProgram.categoryCombo && selectedProgram.categoryCombo.categories.length > 0 && !selectedProgram.categoryCombo.isDefault) {
+                orgUnitSelectorWidth = 3;
+                programSelectorWidth = 5;
+                actionButtonsWidth = 4;
+            } else {
+                orgUnitSelectorWidth = 3;
+                programSelectorWidth = 3;
+                actionButtonsWidth = 6;
+            }
+        } else if (this.props.selectedOrgUnitId && !this.props.selectedProgramId) {
+            orgUnitSelectorWidth = 3;
+            programSelectorWidth = 3;
+            actionButtonsWidth = 6;
+        } 
+
         return (
             <Paper className={this.props.classes.paper}>
-                <Grid container spacing={24}>
-                    <Grid item xs={12} sm={4}>
+                <Grid container spacing={16}>
+                    <Grid item xs={12} sm={orgUnitSelectorWidth}>
                         <OrgUnitSelector
-                            selectedOrgUint={this.props.selectedOrgUnit}
-                            handleChangeOrgUnit={this.handleChangeOrgUnit}
-                            handleClickOrgUnit={this.handleClickOrgUnit} 
+                            selectedOrgUint={this.props.selectedOrgUnitId}
+                            handleClickOrgUnit={this.handleClickOrgUnit}
+                            storedOrgUnits={this.props.storedOrgUnits}
                         />
                     </Grid>
-                    <Grid item xs={12} sm={4}>
+                    <Grid item xs={12} sm={programSelectorWidth}>
                         <ProgramSelector
                             selectedProgram={this.props.selectedProgramId}
-                            handleChangeProgram={this.handleChangeProgram}
+                            selectedCategories={this.props.selectedCategories}
                             handleClickProgram={this.handleClickProgram}
+                            handleSetCatergoryCombo={this.handleSetCatergoryCombo}
+                            resetProgram={this.resetProgram}
```

Comment:
Add parameter for "buttonModeMaxLength" - that controls how many buttons it makes sense to display before switching to list mode.

Reply:
Added parameter.